### PR TITLE
Add asynchronous error injection on edge

### PR DESCRIPTION
**Pull request by larry9523**
Add a capability to inject an asynchronous error and remove all asynchronous errors in zocl driver
1. Add an ioctl to inject/remove error
2. Add an xcl level API to inject/remove error

### CODE REVIEW

File: src/runtime_src/core/include/xrt_error_code.h [hunk 1]
```diff
@@ -102,12 +111,21 @@ enum xrtErrorSeverity {
...
+
+enum xrtErrorModule {
+  XRT_ERROR_MODULE_FIREWALL = 0,
+  XRT_ERROR_MODULE_CMC,
+  XRT_ERROR_MODULE_AIE_CORE,
+  XRT_ERROR_MODULE_AIE_MEMORY,
+  XRT_ERROR_MODULE_AIE_SHIM
+};
+
+enum xrtErorClass {
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_error_code.h:123]: Let fix the typo "Eror" :-)

File: src/runtime_src/core/include/xrt_error_code.h [hunk 2]
```diff
@@ -64,7 +64,25 @@
...
+#define XRT_ERROR_DRIVER_MASK		0xFUL
+#define XRT_ERROR_DRIVER_SHIFT		16
+#define XRT_ERROR_SEVERITY_MASK		0xFUL
+#define XRT_ERROR_SEVERITY_SHIFT	24
+#define XRT_ERROR_MODULE_MASK		0xFUL
+#define XRT_ERROR_MODULE_SHIFT		32
+#define XRT_ERROR_CLASS_MASK		0xFUL
+#define XRT_ERROR_CLASS_SHIFT		40
+
+#define	XRT_BUILD_ERROR_CODE(num, driver, severity, module, eclass) \
```
Comment (stsoe) [inline src/runtime_src/core/include/xrt_error_code.h:79]: While yours read better, I think we need the verb at the end to be consistent.  So let's call it XRT_ERROR_CODE_BUILD.  Part of the reason for this naming convention is that it makes it fairly clear which header declares the API.

File: src/runtime_src/core/include/experimental/xrt-next.h [hunk 3]
```diff
@@ -276,6 +276,35 @@ XCL_DRIVER_DLLESPEC int xclOpenIPInterruptNotify(xclDeviceHandle handle, uint32_
...
+ * @severity:   error severity
+ * @module:     module to inject
+ * @eclass:     error class
+ * Return:      0 or appropriate error number
+ *
+ * Note: this API needs root privilege
+ */
+XCL_DRIVER_DLLESPEC
+int
+xclErrorInject(xclDeviceHandle handle, uint16_t num, uint16_t driver, uint16_t severity, uint16_t module, uint16_t eclass);
```
Comment (stsoe) [inline src/runtime_src/core/include/experimental/xrt-next.h:294]: Who will call this API?

Comment (larry9523) [reply src/runtime_src/core/include/experimental/xrt-next.h:294]: > Who will call this API?

Mainly the developers who wants to test its error handling mechanism while there is no easy way to generate error from hardware, including XRT developer, or application developers, etc.

File: src/runtime_src/core/include/xrt_error_code.h (hunk 2 above)
Comment (larry9523) [reply src/runtime_src/core/include/xrt_error_code.h]: Done.

File: src/runtime_src/core/include/xrt_error_code.h (hunk 1 above)
Comment (larry9523) [reply src/runtime_src/core/include/xrt_error_code.h]: Done

File: src/runtime_src/core/include/experimental/xrt-next.h (hunk 3 above)
Comment (stsoe) [reply src/runtime_src/core/include/experimental/xrt-next.h:294]: Ok.   This is a temporary API then?

**Pull request merged**